---
title: "scATAC-seq Analysis"
author: "Benedict"
date: "02/03/2021"
output:
  pdf_document:
    latex_engine: xelatex
---

# Single Cell ATAC-seq Analysis with Signac
This R Markdown document performs differential accessible chromatin testing for a group of samples

## Project Structure
```
project
│   README.md
└───data
│   │
│   └───RNA
│   │     N2
│   │     B2
│   │     E2
│   │     M2
│   └───ATAC
│   │     X31-OVA_DIV0
│   │     X31-OVA_DIV3
└───code
│     differential_scRNA-seq.Rmd
│     differential_scATAC-seq.Rmd
└───results
│   │
│   └───RNA
│   │   scRNA-seq-DE.xlsx
│   │   expression_heatmap.png
│   │   expression_dotplot.png
│   └───ATAC
│   │     A
│   │     B
```

## Dependencies
* tidyverse
* Seurat
* Signac
* cowplot
* hdf5r
* GenomeInfoDb
* EnsDb.Mmusculus.v79
* patchwork

```{r setup, include=FALSE, echo=TRUE, results='hide'}
setwd(paste0("C:/Users/bened/AppData/Local/Packages/CanonicalGroupLimited.Ubun",
             "tuonWindows_79rhkp1fndgsc/LocalState/rootfs/home/benedict/CaseStu",
             "dy/code"))

# Load Dependencies
packages <- c("tidyverse",  "Seurat", "Signac", "cowplot", "patchwork")
#install.packages(packages)
lapply(packages, library, character.only = TRUE)
BiocManager::install("hdf5r")
library(hdf5r)
library(GenomeInfoDb)
library(EnsDb.Mmusculus.v79)
```

## Data Import
```{r data_import, include=FALSE, echo=TRUE}
data_folder <- "../data/united/ATAC/"
sample_names <- c("X31-OVA_Div0", "X31-OVA_Div3")
```

## Create Seurat objects from input data
### Peak/Cell Matrix
Each row is a region of the genome (peak) predicted to represent a region of
open chromatin. Each value is the number of Tn5 integration sites for each 
cell that map to the peak (row).
### Fragment File
List of all unique fragments across all single cells
```{r construct_objects, include=FALSE, echo=TRUE, results='hide'}
sample_objs <- c()
for(sample_name in sample_names){
  
  # Read in count data
  counts <- Read10X_h5(paste0(data_folder, sample_name, "/filtered_peak_bc_matrix.h5"))
  
  # Read in metadata
  metadata <- read.csv(
    file = paste0(data_folder, sample_name, "/singlecell.csv"),
    header = TRUE,
    row.names = 1
    )
  
  # Create assay
  assay <- CreateChromatinAssay(
    counts = counts,
    sep = c(":", "-"),
    genome = "mm10",
    fragments = paste0(data_folder, sample_name, "/fragments.tsv.gz"),
    min.cells = 1
  )
  
  obj <- CreateSeuratObject(
    counts = assay,
    project = 'ATAC',
    meta.data = metadata,
    assay = 'peaks'
    )
  
  obj@meta.data$state <- sample_name
  sample_objs <- append(sample_objs, obj)
  }
```


```{r annotate, include=FALSE, echo=TRUE}

# extract gene annotations from EnsDb
annotations <- GetGRangesFromEnsDb(ensdb = EnsDb.Mmusculus.v79)

# change to UCSC style since the data was mapped to hg19
seqlevelsStyle(annotations) <- 'UCSC'
genome(annotations) <- "mm10"
```

## Compute QC metrics for each sample
Workflow computes and applied the following commonly used metrics in a QC step:

* Nucleosome banding pattern: Historgram of DNA fragment sizes should correspond
  to the length of DNA wrapped around a nucleosome
* Transcriptional start site enrichment score: poor ATAC-seq experiments will
  typically have a low TSS enrichment score
* Total number of fragments in peaks
  * A measure of cellular sequencing depth
  * Cells with high levels may represent cell multiplets
* Fraction of fragments in peaks
  * Represents fraction of all fragments that fall within ATAC-seq peaks
  * Cells with low values (<15-20%) often indicate the cell is of low-quality
* Ratio of reads in genomic blacklist regions detailed by ENCODE
  * These reads are often associated with artificial signal
  * Cells with a high proportion of reads mapping to areas often
    respresent techincal artifacts and should be removed

```{r compute_qc_metrics, include=FALSE, echo=TRUE, results='hide'}
temp <- c()
for(seurat_obj in sample_objs){
  Annotation(seurat_obj) <- annotations
  seurat_obj <- NucleosomeSignal(object = seurat_obj)
  seurat_obj <- TSSEnrichment(object = seurat_obj, fast = FALSE)
  seurat_obj$pct_reads_in_peaks <- seurat_obj$peak_region_fragments / seurat_obj$passed_filters * 100
  seurat_obj$blacklist_ratio <- seurat_obj$blacklist_region_fragments / seurat_obj$peak_region_fragments
  temp <- append(temp, seurat_obj)
}
sample_objs <- temp
rm(temp)
```


```{r plot_qc_metrics, include=TRUE}
plots <- list()
for (sample_obj in sample_objs){
  plot(VlnPlot(object = sample_obj,
               features = c('peak_region_fragments',
                          'pct_reads_in_peaks',
                          'blacklist_ratio',
                          'nucleosome_signal',
                          'TSS.enrichment'),
               cols = c('light green'),
               combine = TRUE) + 
         plot_annotation(
           title = paste0(sample_obj@meta.data$state[[1]], " QC Metrics")
         )) 
  }
```

```{r results_setup, include=FALSE, echo=TRUE}
if(!dir.exists("../results")){dir.create("../results")}
dir.create("../results/qc_metrics")
```


## Merge Samples
For efficiency, the merged object can be read in instead by uncommenting the second part of this
code chunk. Remember to comment out the first part.
```{r merge_samples, include=FALSE, echo=TRUE}
# Option 1: Perform merge

# Downsample
sample_objs[[-1]] <- sample_objs[[-1]][,1:500]
sample_objs[[1]] <-  sample_objs[[1]][,1:500]

samples.combined <- merge(sample_objs[[1]], y = sample_objs[-1], 
                          add.cell.ids =c("DIV0", "DIV1"), project = "CaseStudy2021")

# Option 2: Read in merged object
#samples.combined <- readRDS("DIV0AND3")
#Idents(samples.combined) <- "state"

```

```{r}
Annotation(samples.combined) <- annotations
```


## Normalize Samples
The data is normalized using term-frequency inverse document frequency
```{r normalise, include=FALSE, echo=TRUE}
samples.combined <- RunTFIDF(samples.combined)
```

## Linear Dimensionality Reduction
```{r dim_red, include=FALSE, echo=TRUE, results='hide'}
samples.combined <- FindTopFeatures(samples.combined, min.cutoff = 'q0')
samples.combined <- RunSVD(samples.combined)
samples.combined <- RunUMAP(object = samples.combined, reduction = 'lsi', dims = 2:30)
```

## UMAP Plot
```{r}
plt <- DimPlot(samples.combined, reduction = "umap", group.by = "state")
plt
#View(samples.combined)
```

## Gene Activity Matrix
```{r construct_gene_activity_matrix, include=FALSE, echo=TRUE}
gene.activities <- GeneActivity(samples.combined)
# add the gene activity matrix to the Seurat object as a new assay and normalize it
samples.combined[['RNA']] <- CreateAssayObject(counts = gene.activities)
samples.combined <- NormalizeData(
  object = samples.combined,
  assay = 'RNA',
  normalization.method = 'LogNormalize',
  scale.factor = median(samples.combined$nCount_RNA)
)
```

```{r}
Assays(samples.combined)
```

```{r}
DefaultAssay(samples.combined) <- 'RNA'

FeaturePlot(
  object = samples.combined,
  features = c('Sell', 'Adam17', 'Klf2', 'Ccr7'),
  pt.size = 0.1,
)
```

## Identify differentially open chromatin regions
For efficiency, the peaks object can be read in instead by uncommenting the second part of this
code chunk. Remember to comment out the first part.
```{r differential_open_chromatin, include=FALSE, echo=TRUE}
DefaultAssay(samples.combined) <- 'peaks'
# Option 1: Perform DE
samples.combined.peaks <- FindAllMarkers(
  object = samples.combined,
  min.pct = 0.01,
  test.use = 'LR',
  latent.vars = 'peak_region_fragments'
)
samples.combined.peaks$regions <- rownames(samples.combined.peaks)
saveRDS(samples.combined.peaks, "peaks_DIV0AND3")

# Option 2: Read in peaks object
# samples.combined.peaks<- readRDS("peaks_DIV0AND3")
```

```{r}
closest_genes <- ClosestFeature(samples.combined,
                                regions = rownames(samples.combined.peaks))
```


### Write to excel spreadsheet 'scATAC-seq-DE'
Results of differential open chromatin regions test are written to an excel workbook with two sheets.
Sheet one is all results with genes of particular importance being highlighted in yellow
Sheet two contains only those genes of particular importance

```{r write_excel, include=FALSE, echo=TRUE}
de_atac_results <- merge(samples.combined.peaks, closest_genes, by.x='regions', by.y='query_region')

excel <- createWorkbook("scATAC-seq-DE")
addWorksheet(excel, "scATAC-seq")
writeData(excel, sheet = 1, de_atac_results)

# Highlight genes of particular interest
genes_of_note <- c("Sell", "Adam17", "Klf2", "Lef1", "Foxo1", "Sp1", "Ets1",
                   "Mzf1", "Irf1", "Ccr7", "S1pr1", "Cd44")
de_atac_results$rownum <- 1:nrow(de_atac_results)
rw <- de_atac_results %>% dplyr::filter(gene_name %in% genes_of_note )
highlight_style <- openxlsx::createStyle(fgFill = "yellow")
openxlsx::addStyle(wb = excel, 
                   sheet = 1, 
                   style = highlight_style,
                   rows = (rw$rownum+1),
                   cols = 1:ncol(rw),
                   stack = TRUE,
                   gridExpand = TRUE)

# Create second sheet containing only genes of note
rw$rownum <- NULL
rw <- rw %>% arrange(gene)
addWorksheet(excel, "scATAC-seq-specific")
writeData(excel, sheet=2, rw)

# Finally write out to file
saveWorkbook(excel, file = "../results/ATAC/scATAC-seq-DE.xlsx", overwrite = TRUE)
```


Edit and run the following chunk if you receive an error when running the
CoveragePlot

```{r include=FALSE, echo=TRUE, results='hide'}
new_paths <- c("C:/Users/bened/AppData/Local/Packages/CanonicalGroupLimited.UbuntuonWindows_79rhkp1fndgsc/LocalState/rootfs/home/benedict/CaseStudy/data/united/ATAC/X31-OVA_Div0/fragments.tsv.gz",
              "C:/Users/bened/AppData/Local/Packages/CanonicalGroupLimited.UbuntuonWindows_79rhkp1fndgsc/LocalState/rootfs/home/benedict/CaseStudy/data/united/ATAC/X31-OVA_Div3/fragments.tsv.gz")
frags <- Fragments(samples.combined)
for (i in 1:length(frags)) {
  frags[[i]] <- UpdatePath(frags[[i]], new.path =  new_paths[[i]])
}
frags
samples.combined <- SetAssayData(samples.combined, slot = "fragments", new.data =frags)
```

## Accessibility Violin Plots
### Sell
```{r violinplot_sell, include=TRUE, echo=TRUE}
Idents(samples.combined) <- "state"
VlnPlot(samples.combined, features = c("Adam17"))
```

### Adam17
```{r violinplot_sell, include=TRUE, echo=TRUE}
Idents(samples.combined) <- "state"
VlnPlot(samples.combined, features = c("Adam17"))
```

### Klf2
```{r violinplot_sell, include=TRUE, echo=TRUE}
Idents(samples.combined) <- "state"
VlnPlot(samples.combined, features = c("Sell"))
```

## Coverage Plots

### Sell
```{r coverage_plot, include=TRUE, echo=FALSE}
DefaultAssay(samples.combined) <- 'peaks'
CoveragePlot(
  object = samples.combined,
  region = "Sell"
)
```

### Adam17
```{r coverage_plot, include=TRUE, echo=FALSE}
CoveragePlot(
  object = samples.combined,
  region = "Adam17"
)
```

### Klf2
```{r coverage_plot, include=TRUE, echo=FALSE}
CoveragePlot(
  object = samples.combined,
  region = "Klf2"
)
```