---
title: "scATAC-seq Analysis"
author: "Benedict"
date: "02/03/2021"
output:
  pdf_document:
    latex_engine: xelatex
---

# Single Cell ATAC-seq Analysis with Signac
This R Markdown document performs differential accessible chromatin testing for a group of samples and consists of six stages:

1. Data Import
2. Quality Control
3. Dimensionality Reduction
4. Miscellaneous plots
5. Differential Accessibility(DA) analysis

## Project Structure
```
project
│   README.md
└───data
│   │
│   └───RNA
│   │     N2
│   │     B2
│   │     E2
│   │     M2
│   └───ATAC
│   │     X31-OVA_DIV0
│   │     X31-OVA_DIV3
└───code
│     differential_scRNA-seq.Rmd
│     differential_scATAC-seq.Rmd
└───results
│   └───RNA
│   │   │ scRNA-seq-DE.xlsx
│   │   └───QC
│   │   │ 
│   │   └───plots
│   └───ATAC
│   │   │  scATAC-seq-DE.xlsx
│   │   └───QC
│   │   │ 
│   │   └───plots
```

## Dependencies
* tidyverse
* Seurat
* Signac
* cowplot
* hdf5r
* GenomeInfoDb
* EnsDb.Mmusculus.v79
* patchwork

```{r setup, include=FALSE, echo=TRUE, results='hide'}
setwd(paste0("C:/Users/bened/AppData/Local/Packages/CanonicalGroupLimited.Ubun",
             "tuonWindows_79rhkp1fndgsc/LocalState/rootfs/home/benedict/CaseStu",
             "dy/code"))

# Load Dependencies
packages <- c("tidyverse",  "Seurat", "Signac", "cowplot", "patchwork", "enrichR")
#install.packages(packages)
lapply(packages, library, character.only = TRUE)
BiocManager::install("hdf5r")
library(hdf5r)
library(GenomeInfoDb)
library(EnsDb.Mmusculus.v79)
```


## Data Import
```{r data_import, include=FALSE, echo=TRUE}
data_folder <- "../data/united/ATAC/"
results_folder <- "../results/ATAC/"
sample_names <- c("naive", "effector", "memory")
```


## Create Seurat objects from input data
### Peak/Cell Matrix
Each row is a region of the genome (peak) predicted to represent a region of
open chromatin. Each value is the number of Tn5 integration sites for each 
cell that map to the peak (row).
### Fragment File
List of all unique fragments across all single cells
```{r construct_objects, include=FALSE, echo=TRUE, results='hide'}
sample_objs <- c()
for(sample_name in sample_names){
  
  # Read in count data
  counts <- Read10X_h5(paste0(data_folder, sample_name, "/filtered_peak_bc_matrix.h5"))
  
  # Read in metadata
  metadata <- read.csv(
    file = paste0(data_folder, sample_name, "/singlecell.csv"),
    header = TRUE,
    row.names = 1
    )
  
  # Create assay
  assay <- CreateChromatinAssay(
    counts = counts,
    sep = c(":", "-"),
    genome = "mm10",
    fragments = paste0(data_folder, sample_name, "/fragments.tsv.gz"),
    min.cells = 1
  )
  
  obj <- CreateSeuratObject(
    counts = assay,
    project = 'ATAC',
    meta.data = metadata,
    assay = 'peaks'
    )
  
  obj@meta.data$state <- sample_name
  sample_objs <- append(sample_objs, obj)
  }
```


Number of cells per sample pre-QC
```{r pre_qc_cell_counts}
for(i in seq_along(sample_objs)){
  num_cells <- ncol(sample_objs[[i]])
  print(paste0(sample_names[[i]],": ", num_cells))
}
```

### Annotations
Add gene annotations to the seruat objects
```{r annotate, include=FALSE, echo=TRUE}

# extract gene annotations from EnsDb
annotations <- GetGRangesFromEnsDb(ensdb = EnsDb.Mmusculus.v79)

seqlevelsStyle(annotations) <- 'UCSC'
genome(annotations) <- "mm10"
```

## Compute QC metrics for each sample
Workflow computes and applied the following commonly used metrics in a QC step:

* Nucleosome banding pattern: Historgram of DNA fragment sizes should correspond
  to the length of DNA wrapped around a nucleosome
* Transcriptional start site enrichment score: poor ATAC-seq experiments will
  typically have a low TSS enrichment score
* Total number of fragments in peaks
  * A measure of cellular sequencing depth
  * Cells with high levels may represent cell multiplets
* Fraction of fragments in peaks
  * Represents fraction of all fragments that fall within ATAC-seq peaks
  * Cells with low values (<15-20%) often indicate the cell is of low-quality
* Ratio of reads in genomic blacklist regions detailed by ENCODE
  * These reads are often associated with artificial signal
  * Cells with a high proportion of reads mapping to areas often
    respresent techincal artifacts and should be removed

```{r compute_qc_metrics, include=FALSE, echo=TRUE, results='hide'}
temp <- c()
for(seurat_obj in sample_objs){
  Annotation(seurat_obj) <- annotations
  seurat_obj <- NucleosomeSignal(object = seurat_obj)
  seurat_obj <- TSSEnrichment(object = seurat_obj, fast = FALSE)
  seurat_obj$pct_reads_in_peaks <- seurat_obj$peak_region_fragments / seurat_obj$passed_filters * 100
  seurat_obj$blacklist_ratio <- seurat_obj$blacklist_region_fragments / seurat_obj$peak_region_fragments
  temp <- append(temp, seurat_obj)
}
sample_objs <- temp
rm(temp)
```


```{r plot_qc_metrics, include=TRUE}
qc_location <- paste0(results_folder, "QC/")
# Temp merge for plot purposes only
#samples.combined <- merge_samples(precomputed = TRUE, filename = "N-E-M-complete")

plt_num_frags_in_peaks <- VlnPlot(object = samples.combined, 
                                  features = c("peak_region_fragments"),
                                  group.by = "state") +  labs(title = "Number of Fragments in Peaks")
save_plot(paste0(qc_location, "num_frags_in_peaks", ".png"), plt_num_frags_in_peaks)


plt_pct_reads_in_peaks <- VlnPlot(object = samples.combined,
                                  features = c("pct_reads_in_peaks"),
                                  group.by = "state") +  labs(title = "% Reads In Peaks")
save_plot(paste0(qc_location, "pct_reads_in_peaks", ".png"), plt_pct_reads_in_peaks)


plt_blacklist_ratio <- VlnPlot(object = samples.combined,
                               features = c("blacklist_ratio"),
                               group.by = "state") +  labs(title = "Blacklist Ratio")
save_plot(paste0(qc_location, "blacklist_ratio", ".png"), plt_blacklist_ratio)


plt_nucleosome_signal <- VlnPlot(object = samples.combined,
                                 features = c("nucleosome_signal"),
                                 group.by = "state") +  labs(title = "Nucleosome Signal")
save_plot(paste0(qc_location, "nucleosome_signal", ".png"), plt_nucleosome_signal)

plt_tss_enrichment <- VlnPlot(object = samples.combined, 
                              features = c("TSS.enrichment"),
                              group.by = "state") + labs(title = "TSS Enrichment")
save_plot(paste0(qc_location, "tss_enrichment", ".png"), plt_tss_enrichment)
```


```{r}
samples.combined <- subset(
  x = samples.combined,
  subset = peak_region_fragments > 3000 &
    peak_region_fragments < 50000 &
    pct_reads_in_peaks > 20 &
    blacklist_ratio < 0.025 &
    nucleosome_signal < 4 &
    TSS.enrichment > 2
)
```


Number of cells per sample post-QC
```{r post_qc_cell_counts}
for(i in seq_along(sample_objs)){
  num_cells <- ncol(sample_objs[[i]])
  print(paste0(names[[i]],": ", num_cells))
}
```


```{r}
samples.combined <- merge_samples(precomputed = TRUE, filename = "N-E-M-complete")
```


```{r results_setup, include=FALSE, echo=TRUE}
if(!dir.exists("../results")){dir.create("../results")}
dir.create("../results/qc_metrics")
```


## Merge Samples
For efficiency, the merged object can be read in instead by uncommenting the second part of this
code chunk. Remember to comment out the first part.
```{r merge_samples, include=FALSE, echo=TRUE}

merge_samples <- function(precomputed=FALSE, filename=NULL){
  if(precomputed == TRUE){
    # Option 1: Perform merge
    samples.combined <- readRDS(paste0(data_folder, "PRE-COMPUTED/", filename))
  }
  else{
    # Option 2: Read in merged object
    samples.combined <- merge(sample_objs[[1]], 
                              y = sample_objs[-1], 
                              add.cell.ids = sample_names,
                              project = "scATAC-seq")
    saveRDS(samples.combined, file=paste0(data_folder, "PRE-COMPUTED/", filename))
  }
  Idents(samples.combined) <- "state"
  return(samples.combined)
}
```


```{r}
Annotation(samples.combined) <- annotations
```


## Normalize Samples
The data is normalized using term-frequency inverse document frequency
```{r normalise, include=FALSE, echo=TRUE}
samples.combined <- RunTFIDF(samples.combined)
```


## Linear Dimensionality Reduction
```{r dim_red, include=FALSE, echo=TRUE, results='hide'}
samples.combined <- FindTopFeatures(samples.combined)
samples.combined <- RunSVD(samples.combined)
samples.combined <- RunUMAP(object = samples.combined, reduction = 'lsi', dims = 2:30)
```


## UMAP Plot
```{r}
samples.combined <- FindNeighbors(object = samples.combined, reduction = 'lsi', dims = 2:30)
samples.combined <- FindClusters(object = samples.combined, verbose = FALSE, algorithm = 3, resolution = 0.2)
plt <- DimPlot(samples.combined, reduction = "umap", group.by = "state")
plt_2 <- DimPlot(samples.combined, reduction = "umap")
plt
plt_2
```

## Gene Activity Matrix
```{r construct_gene_activity_matrix, include=FALSE, echo=TRUE}
gene.activities <- GeneActivity(samples.combined)
# add the gene activity matrix to the Seurat object as a new assay and normalize it
samples.combined[['RNA']] <- CreateAssayObject(counts = gene.activities)
samples.combined <- NormalizeData(
  object = samples.combined,
  assay = 'RNA',
  normalization.method = 'LogNormalize',
  scale.factor = median(samples.combined$nCount_RNA)
)
```


```{r sell_feature_plot, include=TRUE}
DefaultAssay(samples.combined) <- 'RNA'

plt <- FeaturePlot(
  object = samples.combined,
  features = c('Sell'),
  pt.size = 0.7,
)
save_plot("../results/ATAC/plots/sell_RNA_activity.png", plot=plt)
plt
```

```{r adam17_feature_plot, include=TRUE}
plt <- FeaturePlot(
 object = samples.combined,
  features = c('Adam17'),
  pt.size = 0.7,
)
save_plot("../results/ATAC/plots/adam17_RNA_activity.png", plot=plt)
plt
```

```{r klf2_feature_plot, include=TRUE}
plt <- FeaturePlot(
 object = samples.combined,
  features = c('Klf2'),
  pt.size = 0.7,
)
save_plot("../results/ATAC/plots/klf2_RNA_activity.png", plot=plt)
plt
```

## Coverage Plot
```{r coverage_plot, include=TRUE, echo=FALSE}
plt <- CoveragePlot(object = samples.combined,region = "Sell")
save_plot("../results/ATAC/plots/coverage.png", plot=plt)
plt
```

## Identify differentially open chromatin regions
For efficiency, the peaks object can be read in instead by uncommenting the second part of this
code chunk. Remember to comment out the first part.

Edit and run the following chunk if you receive an error when running the
CoveragePlot

```{r include=FALSE, echo=TRUE, results='hide'}
new_paths <- c("C:/Users/bened/AppData/Local/Packages/CanonicalGroupLimited.UbuntuonWindows_79rhkp1fndgsc/LocalState/rootfs/home/benedict/CaseStudy/data/united/ATAC/X31-OVA_Div0/fragments.tsv.gz",
              "C:/Users/bened/AppData/Local/Packages/CanonicalGroupLimited.UbuntuonWindows_79rhkp1fndgsc/LocalState/rootfs/home/benedict/CaseStudy/data/united/ATAC/X31-OVA_Div3/fragments.tsv.gz")
frags <- Fragments(samples.combined)
for (i in 1:length(frags)) {
  frags[[i]] <- UpdatePath(frags[[i]], new.path =  new_paths[[i]])
}
frags
samples.combined <- SetAssayData(samples.combined, slot = "fragments", new.data =frags)
```


```{r differential_accessibility, include=FALSE, echo=TRUE, results='hide'}
DefaultAssay(samples.combined) <- 'peaks'
Idents(samples.combined) <- "state"
genes <- c("Sell", "Adam17", "Klf2", "Lef1", "Foxo1", "Sp1", "Ets1", "Mzf1",
           "Irf1", "Ccr7", "S1pr1", "Cd44", "Ccl5")
res <- diff_acc(sample_a = "naive", sample_b = "effector") %>% arrange(p_val_adj)
write_to_xlsx(data = res, 
              filename = "scATAC-NaivevsEffectorseq", 
              genes_of_interest = genes)
```


```{r}
diff_acc <- function(sample_a, sample_b){
  su <- subset(samples.combined, idents = c(sample_a, sample_b))

  CD8_difference <- Seurat::FindMarkers(object = su, 
                                        test.use = 'LR',
                                        ident.1=sample_a,
                                        ident.2=sample_b,
                                        min.pct=0.2,
                                        latent.vars = 'peak_region_fragments'
                                        )

  query_region <- rownames(CD8_difference)
  CD8_difference$query_region <- query_region
  closest_genes <- ClosestFeature(su, 
                                  regions = query_region)
  
  res <- merge(CD8_difference, closest_genes, by=c("query_region"))

  return(res) 
}
```


### Write to excel spreadsheet 'scATAC-seq-DE'
Results of differential open chromatin regions test are written to an excel workbook with two sheets. Sheet one is all results with genes of particular importance being highlighted in yellow. Sheet two contains only those genes of particular importance.

**avg_log2fc:** log fold-change in average expression between the two groups
**pct.1:** Percentage of cells where gene is detected in first group
**pct.2:** Percentage of cells where gene is detected in second group
**p_val_adj:** Adjusted p-value using bonferroni correction using all genes in dataset
```{r write_excel, include=TRUE, echo=TRUE}
write_to_xlsx <- function(data, filename, genes_of_interest){
  excel <- createWorkbook(filename)
  addWorksheet(excel, "all_genes")
  writeData(excel, sheet = 1, data)
  
  # Highlight genes of particular interest
  data$rownum <- 1:nrow(data)
  rw <- data %>% dplyr::filter(gene_name %in% genes_of_interest )
  highlight_style <- openxlsx::createStyle(fgFill = "yellow")
  openxlsx::addStyle(wb = excel, 
                     sheet = 1, 
                     style = highlight_style,
                     rows = (rw$rownum+1),
                     cols = 1:ncol(rw),
                     stack = TRUE,
                     gridExpand = TRUE)
  
  # Create second sheet containing only genes of note
  rw$rownum <- NULL
  rw <- rw %>% arrange(gene_name)
  addWorksheet(excel, "genes_of_interest")
  writeData(excel, sheet=2, rw)
  
  
  # Gene Enrichment Analysis
    # Find and plot top enriched and top depleted GO terms from EnrichR
  go_terms <- enrichr(genes = unique(res$gene_name),
                      databases = "KEGG_2019_Mouse")[["KEGG_2019_Mouse"]] 
  go_terms <- go_terms[c("Term", "Overlap", "Adjusted.P.value", "Genes")] %>%
    arrange("Overlap")
  addWorksheet(excel, "go_terms")
  writeData(excel, sheet=3, go_terms)
  
  # Finally write out to file
  saveWorkbook(excel, file = paste0("../results/ATAC/", filename, ".xlsx"),
               overwrite = TRUE)
  }
```
