---
title: "scRNA-seq Analysis"
author: "Benedict"
date: "03/03/2021"
output:
  pdf_document:
    latex_engine: xelatex
---

# Single Cell RNA-seq Analysis with Seurat
This R Markdown document performs differential gene expression testing for a group of samples and consists of six stages:

1. Data Import
2. Quality Control
3. Dimensionality Reduction
4. Miscellaneous plots
5. Differential Expression (DE) analysis
6. Gene Enrichment Analysis

## Project Structure
```
project
│   README.md
└───data
│   │
│   └───RNA
│   │     N2
│   │     B2
│   │     E2
│   │     M2
│   └───ATAC
│   │     X31-OVA_DIV0
│   │     X31-OVA_DIV3
└───code
│     differential_scRNA-seq.Rmd
│     differential_scATAC-seq.Rmd
└───results
│   └───RNA
│   │     scRNA-seq-DE.xlsx
│   │     expression_heatmap.png
│   │     expression_dotplot.png
│   └───ATAC
│         scATAC-seq-DE.xlsx
```

## Dependencies
* tidyverse
* Seurat
* cowplot
* biomart
* openxlsx
* patchwork
* enrichR

```{r setup, include=FALSE, echo=TRUE, results='hide'}
knitr::opts_chunk$set(echo = TRUE)
setwd(paste0("C:/Users/bened/AppData/Local/Packages/CanonicalGroupLimited.Ubun",
             "tuonWindows_79rhkp1fndgsc/LocalState/rootfs/home/benedict/CaseStu",
             "dy/code"))

# Load Dependencies
packages <- c("tidyverse",  "Seurat", "cowplot", "biomaRt", "openxlsx",
              "patchwork", "enrichR")
#install.packages(packages)
lapply(packages, library, character.only = TRUE)
```

## Data Import
Uncomment following code block to analyse the DIV0-4 samples
```{r data_import_a,  include=TRUE, echo=TRUE, results='hide'}
# data_folder <- "../data/united/RNA/"
# 
# div_0.data <-  Read10X(data.dir = paste0(data_folder,
#                                          "X31-OVA-D0/filtered_feature_bc_matrix"))
# div_1.data <-  Read10X(data.dir = paste0(data_folder,
#                                          "X31-OVA-D1/filtered_feature_bc_matrix"))
# div_2.data <-  Read10X(data.dir = paste0(data_folder,
#                                          "X31-OVA-D2/filtered_feature_bc_matrix"))
# div_3.data <-  Read10X(data.dir = paste0(data_folder,
#                                          "X31-OVA-D3/filtered_feature_bc_matrix"))
# div_4.data <-  Read10X(data.dir = paste0(data_folder,
#                                          "X31-OVA-D4/filtered_feature_bc_matrix"))
# samples <- c(div_0.data, div_1.data, div_2.data, div_3.data, div_4.data)
# names <- c("DIV0", "DIV1", "DIV2", "DIV3", "DIV4")
```

Uncomment following code block to analyse the N2, B2, E2, M2 samples
```{r data_import_b, include=TRUE, echo=TRUE}
data_folder <- "../data/united/RNA/"
N2.data <- Read10X(data.dir = paste0(data_folder,
                                     "N2/filtered_feature_bc_matrix"))
B2.data <- Read10X(data.dir = paste0(data_folder,
                                     "B2/filtered_feature_bc_matrix"))
E2.data <- Read10X(data.dir = paste0(data_folder,
                                     "E2/filtered_feature_bc_matrix"))
M2.data <- Read10X(data.dir = paste0(data_folder,
                                     "M2/filtered_feature_bc_matrix"))
samples <- c(N2.data, B2.data, E2.data, M2.data)
names <- c("N2", "B2", "E2", "M2")
```

## Create Seurat objects from input data
```{r create_seurat_objects, include=TRUE, echo=TRUE}
sample_objs <- c()
idx = 1
for(sample in samples){
 sample <- CreateSeuratObject(counts = sample,
                     project = "CaseStudy2021",
                     min.cells = 3,
                     min.features = 200)
 sample@meta.data$state <- names[idx]
 sample_objs <- append(sample_objs, sample)
 idx = idx + 1
}
```

\newpage

## Compute QC metrics

Workflow computes and applied the following commonly used metrics in a QC step:

* Number of unique genes detected in each cell
  * Low-quality cells or empty droplets often have very few genes
  * Cell multiplets may have an abnormally high gene count
* Total number of molecules detected within a cell (should correlate with first metric)
*  Percentage of reads mapping to the mitochondrial genome
  * Low-quality and dying cells often contain high mitochondrial contamination
  
```{r compute_qc_metrics, include=TRUE, echo=TRUE, results='hide'}
temp <- c()
for(seurat_obj in sample_objs){
  total_counts_per_cell <- colSums(seurat_obj@assays$RNA@counts)
  
  mito.genes <- rownames(seurat_obj)[grep("^mt-", rownames(seurat_obj))]
  seurat_obj$percent.mito <- colSums(seurat_obj@assays$RNA@counts[mito.genes, ])/total_counts_per_cell
  
  ribo.genes <- rownames(seurat_obj)[grep("^Rp[sl]", rownames(seurat_obj))]
  seurat_obj$percent.ribo <- colSums(seurat_obj@assays$RNA@counts[ribo.genes, ])/total_counts_per_cell
  
  temp <- append(temp, seurat_obj)
}
sample_objs <- temp
rm(temp)
```


```{r plot_qc_metrics, include=TRUE}
plots <- list()
for (sample_obj in sample_objs){
  plot(VlnPlot(object = sample_obj,
             features = c("percent.ribo", "percent.mito"),
             cols = c('light green'),
             combine = TRUE) + 
         plot_annotation(
           title = paste0(sample_obj@meta.data$state[[1]], " QC Metrics")
         )) 
  }
```
```{r}

# Compute the relative expression of each gene per cell Use sparse matrix
# operations, if your dataset is large, doing matrix devisions the regular way
# will take a very long time.
# for (sample_obj in sample_objs){
# data.filt <- sample_obj
# par(mar = c(4, 8, 2, 1))
# C <- data.filt@assays$RNA@counts
# C <- Matrix::t(Matrix::t(C)/Matrix::colSums(C)) * 100
# most_expressed <- order(apply(C, 1, median), decreasing = T)[20:1]
# boxplot(as.matrix(t(C[most_expressed, ])), cex = 0.1, las = 1, xlab = "% total count per cell", 
#     col = (scales::hue_pal())(20)[20:1], horizontal = TRUE)
# }
```
## Filter Genes
Mitochondrial and ribosomal genes are removed
Comment out this code block to keep them
```{r filter_genes, include=FALSE, echo=TRUE}
temp <- c()
for (sample_obj in sample_objs){
# Filter MALAT1
sample_obj <- sample_obj[!grepl("Malat1", rownames(sample_obj)), ]

# Filter Mitocondrial
sample_obj <- sample_obj[!grepl("^mt-", rownames(sample_obj)), ]

# Filter Ribossomal gene (optional if that is a problem on your data) data.filt
sample_obj<- sample_obj[ ! grepl('^Rp[sl]', rownames(sample_obj)), ]
temp <- append(temp, sample_obj)
}

sample_objs <- temp
rm(temp)
```

## Filter cells based on QC metrics
```{r filter_on_qc_metrics, echo=FALSE}
sample_objs <- lapply(X = sample_objs, FUN = function(x) {
  x <- subset(x = x,
              subset = nFeature_RNA > 200 & nFeature_RNA < 8000 & percent.mito < 10 )
})
```

## Normalize Samples
The data is normalized using a global-scaling normalization method to normalise gene expression levels for each cell by the total expression.
This is multipliplied the a scale factor of 1000 and the result is natural-log-transformed.
```{r normalise, include=FALSE, echo=TRUE}
sample_objs <- lapply(X = sample_objs, FUN = function(x) {
    x <- NormalizeData(x, normalization.method="LogNormalize", scale.factor=1000)
    x <- FindVariableFeatures(x, selection.method = "vst", nfeatures = 2000)
})
```

## Merge Samples
```{r merge_objects, include=TRUE, echo=TRUE}
samples.combined <- merge(sample_objs[[1]], y = sample_objs[-1],
                         add.cell.ids =names, project = "CaseStudy2021")
```

## Scale Data
Pre-processing step before dimensional reduction is performed

* Expression of each gene shifted to give mean expression of 0 across cells
* Expression of each gene scaled to give a variance of 1 across cells
  * Prevents domination from highly-expressed genes
```{r include=TRUE, include=FALSE, echo=TRUE, results='hide'}
samples.combined <- FindVariableFeatures(samples.combined,
                                         selection.method = "vst",
                                         nfeatures = 2000)
all.genes <- rownames(samples.combined)
samples.combined <- ScaleData(samples.combined, features = all.genes)
```

## Linear Dimensionality Reduction
```{r dim_red, include=FALSE, echo=TRUE, results='hide'}
samples.combined <- RunPCA(samples.combined, npcs = 30, verbose = FALSE)
samples.combined <- RunUMAP(samples.combined, reduction = "pca", dims = 2:30)
```

## UMAP Plot
```{r}
plt <- DimPlot(samples.combined, reduction = "umap", group.by = "state")
save_plot("../results/RNA/plots/umap.png", plt)
plt
```
\newpage

```{r}
plt <- DimPlot(samples.combined, reduction = "umap", group.by = "state")
save_plot("../results/RNA/plots/umap_sell.png", plt)
plt
```
\newpage

```{r}
plt <- FeaturePlot(samples.combined, features = c("Adam17"))
save_plot("../results/RNA/plots/umap_adam17.png", plt)
plt
```
\newpage

```{r}
plt <- FeaturePlot(samples.combined, features = c("Klf2"))
save_plot("../results/RNA/plots/umap_klf2.png", plt)
plt
```

## Expression Violin Plots
### Sell
```{r violinplot_sell, include=TRUE, echo=TRUE}
Idents(samples.combined) <- "state"
plt <- VlnPlot(samples.combined, features = c("Sell"))
save_plot("../results/RNA/plots/sell_violin.png", plt)
plt
```

### Adam17

```{r violinplot_adam17, include=TRUE, echo=TRUE}
plt <- VlnPlot(samples.combined, features = c("Adam17"))
save_plot("../results/RNA/plots/adam17_violin.png", plt)
plt
```

### Klf2

```{r violinplot_klf2, include=TRUE, echo=TRUE}
plt <- VlnPlot(samples.combined, features = c("Klf2"))
save_plot("../results/RNA/plots/adam17_violin.png", plt)
plt
```
\newpage

## Expression Dot Plot
```{r dotplot, include=TRUE, echo=TRUE}
genes_of_note <- c("Sell", "Adam17", "Klf2", "Lef1", "Foxo1", "Sp1", "Ets1",
                   "Mzf1", "Irf1", "Ccr7", "S1pr1", "Cd44")
plt <- DotPlot(samples.combined, features = rev(genes_of_note), dot.scale = 8) +
  RotatedAxis()
save_plot("../results/RNA/plots/expression_dotplot.png", plt)
plt
```

\newpage

## Expression Heatmap
```{r heatmap, include=TRUE, echo=TRUE}
samples.combined <- ScaleData(samples.combined)
plt <- DoHeatmap(samples.combined, features=genes_of_note)
save_plot("../results/RNA/plots/expression_heatmap.png", plt)
plt
```

\newpage

## Perform Differential Expression (DE) Testing

### Compare expression betwen markers
Performs a Wilcoxon Rank Sum test to identify differentially expressed genes between a sample and all other cells, for each sample

```{r differential_expression, include=FALSE, echo=TRUE, results='hide'}
Idents(object = samples.combined) <- samples.combined@meta.data$state
samples.combined.markers <- Seurat::FindAllMarkers(object = samples.combined, 
                                                   logfc.threshold = 0.01)
```

### Annotate DE genes from ensembl
```{r annotation_biomart, include=FALSE, echo=TRUE}
ensembl <- useEnsembl(biomart = "ensembl", dataset = "mmusculus_gene_ensembl")
genes <- unique(samples.combined.markers$gene)
genedesc <- getBM(attributes=c('external_gene_name',
                               'description',
                               'chromosome_name',
                               'start_position',
                               'end_position'), 
                  filters = 'external_gene_name', 
                  values = genes,
                  mart =ensembl)

samples.combined.markers <- merge(samples.combined.markers, genedesc, all=TRUE,
                                 by.x = c("gene"), by.y = c("external_gene_name"))

# Reorder columns
new_order <- c("gene", "cluster", "avg_log2FC", "p_val_adj", "pct.1", "pct.2",
               "description", "chromosome_name", "start_position", "end_position")
samples.combined.markers <- samples.combined.markers[new_order]
```

### Write to excel spreadsheet 'scRNA-seq-DE'
Results of differential gene expression test are written to an excel workbook with two sheets.
Sheet one is all results with genes of particular importance being highlighted in yellow
Sheet two contains only those genes of particular importance
```{r write_excel, include=TRUE, echo=TRUE}

excel <- createWorkbook("scRNA-seq-DE")
addWorksheet(excel, "scRNA-seq")
writeData(excel, sheet = 1, samples.combined.markers)

# Highlight genes of particular interest
samples.combined.markers$rownum <- 1:nrow(samples.combined.markers)
rw <- samples.combined.markers %>% dplyr::filter(gene %in% genes_of_note )
highlight_style <- openxlsx::createStyle(fgFill = "yellow")
openxlsx::addStyle(wb = excel, 
                   sheet = 1, 
                   style = highlight_style,
                   rows = (rw$rownum+1),
                   cols = 1:ncol(rw),
                   stack = TRUE,
                   gridExpand = TRUE)

# Create second sheet containing only genes of note
rw$rownum <- NULL
rw <- rw %>% arrange(gene)
addWorksheet(excel, "scRNA-seq-specific")
writeData(excel, sheet=2, rw)

# Finally write out to file
saveWorkbook(excel, file = "../results/RNA/scRNA-seq-DE.xlsx", overwrite = TRUE)
```


## Gene Enrichment Analysis
Find and plot top enriched and top depleted GO terms from EnrichR
```{r gene_ontology_enrichment_analysis, include=TRUE, echo=FALSE}
plt <- DEenrichRPlot(samples.combined, enrich.database = "KEGG_2019_Mouse",
                     max.genes = 300, ident.1 = "N2",
                     ident.2 = "B2", balanced=TRUE, num.pathway = 15)
save_plot("../results/RNA/plots/enrichment_analysis.png", plt)
plt
```
