---
title: "scATAC-seq Analysis"
author: "Benedict"
date: "02/03/2021"
output:
  html_document:
    df_print: paged
  word_document: default
  pdf_document: default
---

# Single Cell ATAC-seq Analysis with Signac
This R Markdown document performs differential accessible chromatin testing for a group of samples

## Project Structure
```
project
│   README.md
└───data
│   │
│   └───RNA
│   │     N2
│   │     B2
│   │     E2
│   │     M2
│   └───ATAC
│   │     X31-OVA_DIV0
│   │     X31-OVA_DIV3
└───code
│     differential_scRNA-seq.Rmd
│     differential_scATAC-seq.Rmd
└───results
│   │
│   └───RNA
│   │   scRNA-seq-DE.xlsx
│   │   expression_heatmap.png
│   │   expression_dotplot.png
│   └───ATAC
│   │     A
│   │     B
```

## Dependencies
* tidyverse
* Seurat
* Signac
* cowplot
* hdf5r
* GenomeInfoDb
* EnsDb.Mmusculus.v79
* patchwork

```{r setup, include=TRUE, echo=TRUE, results='hide'}
setwd(paste0("C:/Users/bened/AppData/Local/Packages/CanonicalGroupLimited.Ubun",
             "tuonWindows_79rhkp1fndgsc/LocalState/rootfs/home/benedict/CaseStu",
             "dy/code"))

# Load Dependencies
packages <- c("tidyverse",  "Seurat", "Signac", "cowplot", "patchwork")
#install.packages(packages)
lapply(packages, library, character.only = TRUE)
BiocManager::install("hdf5r")
library(hdf5r)
library(GenomeInfoDb)
library(EnsDb.Mmusculus.v79)
```

## Data Import
```{r data_import, include=TRUE, echo=TRUE}
data_folder <- "../data/united/ATAC/"
sample_names <- c("X31-OVA_Div0", "X31-OVA_Div3")
```

## Create Seurat objects from input data
```{r construct_objects, include=TRUE, echo=TRUE, results='hide'}
sample_objs <- c()
for(sample_name in sample_names){
  
  # Read in count data
  counts <- Read10X_h5(paste0(data_folder, sample_name, "/filtered_peak_bc_matrix.h5"))
  
  # Read in metadata
  metadata <- read.csv(
  file = paste0(data_folder, sample_name, "/singlecell.csv"),
  header = TRUE,
  row.names = 1
  )
  
  # Create assay
  assay <- CreateChromatinAssay(
    counts = counts,
    sep = c(":", "-"),
    genome = "mm10",
    fragments = paste0(data_folder, sample_name, "/fragments.tsv.gz"),
    min.cells = 1
  )
  
  obj <- CreateSeuratObject(
    counts = assay,
    project = 'ATAC',
    meta.data = metadata
    )
  
  obj@meta.data$state <- sample_name
  sample_objs <- append(sample_objs, obj)
  }
```

```{r annotate, include=TRUE, echo=TRUE}

# extract gene annotations from EnsDb
annotations <- GetGRangesFromEnsDb(ensdb = EnsDb.Mmusculus.v79)

# change to UCSC style since the data was mapped to hg19
seqlevelsStyle(annotations) <- 'UCSC'
genome(annotations) <- "mm10"
```

## Compute QC metrics for each sample
```{r compute_qc_metrics, include=TRUE, echo=TRUE, results='hide'}
temp <- c()
for(seurat_obj in sample_objs){
  Annotation(seurat_obj) <- annotations
  seurat_obj <- NucleosomeSignal(object = seurat_obj)
  seurat_obj <- TSSEnrichment(object = seurat_obj, fast = FALSE)
  seurat_obj$pct_reads_in_peaks <- seurat_obj$peak_region_fragments / seurat_obj$passed_filters * 100
  seurat_obj$blacklist_ratio <- seurat_obj$blacklist_region_fragments / seurat_obj$peak_region_fragments
  temp <- append(temp, seurat_obj)
}
sample_objs <- temp
rm(temp)
```


```{r results_setup, include=TRUE, echo=TRUE}
if(!dir.exists("../results")){dir.create("../results")}
dir.create("../results/qc_metrics")
```


```{r plot_qc_metrics, include=TRUE, echo=TRUE}
plots <- list()
for (seurat_obj in sample_objs){
  plt <- VlnPlot(object = seurat_obj,
                 features = c('peak_region_fragments',
                              'pct_reads_in_peaks',
                              'blacklist_ratio',
                              'nucleosome_signal',
                              'TSS.enrichment'),
                 pt.size = 0.1,
                 ncol = 5,
                 fill.by = 'feature',
                 combine = FALSE)
  
  plots <- append(plots, plt)
}
plt <- wrap_plots(plots = plots, nrow = length(sample_objs), ncol = 5)
save_plot("../results/qc_metrics/qc_metrics_atac.png", plt, base_height = 15, base_width = 20)
plt
```


## Merge Samples
For efficiency, the merged object can be read in instead by uncommenting the second part of this
code chunk. Remember to comment out the first part.
```{r merge_samples, include=TRUE, echo=TRUE}
# Option 1: Perform merge
#samples.combined <- merge(sample_objs[[1]], y = sample_objs[-1], 
#                          add.cell.ids =names, project = "CaseStudy2021")

# Option 2: Read in merged object
samples.combined <- readRDS("DIV0AND3")
Idents(samples.combined) <- "state"

```


## Identify differentially open chromatin regions
For efficiency, the peaks object can be read in instead by uncommenting the second part of this
code chunk. Remember to comment out the first part.
```{r differential_open_chromatin, include=TRUE, echo=TRUE}
# Option 1: Perform DE
# samples.combined.peaks <- FindAllMarkers(
#   object = samples.combined,
#   min.pct = 0.2,
#   test.use = 'LR',
#   latent.vars = 'peak_region_fragments'
# )
# samples.combined.peaks$regions <- rownames(samples.combined.peaks)
# saveRDS(samples.combined.peaks, "peaks_DIV0AND3")

# Option 2: Read in peaks object
samples.combined.peaks<- readRDS("peaks_DIV0AND3")
closest_genes <- ClosestFeature(samples.combined,
                                regions = rownames(samples.combined.peaks))
```

Edit and run the following chunk if you receive an error when running the
CoveragePlot

```{r coverage_plot, include=TRUE, echo=TRUE, results='hide'}
new_paths <- c("C:/Users/bened/AppData/Local/Packages/CanonicalGroupLimited.UbuntuonWindows_79rhkp1fndgsc/LocalState/rootfs/home/benedict/CaseStudy/data/united/ATAC/X31-OVA_Div0/fragments.tsv.gz",
              "C:/Users/bened/AppData/Local/Packages/CanonicalGroupLimited.UbuntuonWindows_79rhkp1fndgsc/LocalState/rootfs/home/benedict/CaseStudy/data/united/ATAC/X31-OVA_Div3/fragments.tsv.gz")
frags <- Fragments(samples.combined)
for (i in 1:length(frags)) {
  frags[[i]] <- UpdatePath(frags[[i]], new.path =  new_paths[[i]])
}
frags
samples.combined <- SetAssayData(samples.combined, slot = "fragments", new.data =frags)
```

```{r coverage_plot, include=FALSE, echo=FALSE}
CoveragePlot(
  object = samples.combined,
  region = "Sell"
)
```