---
title: "Differential scRNA-seq"
author: "Benedict"
date: "18/02/2021"
output: pdf_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
setwd(paste0("C:/Users/bened/AppData/Local/Packages/CanonicalGroupLimited.Ubun",
             "tuonWindows_79rhkp1fndgsc/LocalState/rootfs/home/benedict/CaseStu",
             "dy/code"))

# Load Dependencies
packages <- c("tidyverse",  "Seurat", "cowplot", "biomaRt", "openxlsx")
install.packages(packages)
lapply(packages, library, character.only = TRUE)
```


```{r data_import}
data_folder <- "../data/united/"

# div_0.data <-  Read10X(data.dir = paste0(data_folder, 
#                                          "X31-OVA-D0/filtered_feature_bc_matrix"))
# div_1.data <-  Read10X(data.dir = paste0(data_folder, 
#                                          "X31-OVA-D1/filtered_feature_bc_matrix"))
# div_2.data <-  Read10X(data.dir = paste0(data_folder,
#                                          "X31-OVA-D2/filtered_feature_bc_matrix"))
# div_3.data <-  Read10X(data.dir = paste0(data_folder,
#                                          "X31-OVA-D3/filtered_feature_bc_matrix"))
# div_4.data <-  Read10X(data.dir = paste0(data_folder,
#                                          "X31-OVA-D4/filtered_feature_bc_matrix"))


N2.data <- Read10X(data.dir = paste0(data_folder,
                                     "N2/filtered_feature_bc_matrix"))
B2.data <- Read10X(data.dir = paste0(data_folder, 
                                     "B2/filtered_feature_bc_matrix"))
E2.data <- Read10X(data.dir = paste0(data_folder, 
                                     "E2/filtered_feature_bc_matrix"))
M2.data <- Read10X(data.dir = paste0(data_folder, 
                                     "M2/filtered_feature_bc_matrix"))

samples <- c(N2.data, B2.data, E2.data, M2.data)
names <- c("N2", "B2", "E2", "M2")
```


```{r create_seurat_objects, echo=FALSE}
sample_objs <- c()
idx = 1
for(sample in samples){
 sample <- CreateSeuratObject(counts = sample,
                     project = "CaseStudy2021",
                     min.cells = 3,
                     min.features = 200)
 sample@meta.data$state <- names[idx]
 sample_objs <- append(sample_objs, sample)
 idx = idx + 1
}
```

```{r normalise}
sample_objs <- lapply(X = sample_objs, FUN = function(x) {
    x <- NormalizeData(x)
})
```

```{r merge_objects, echo=FALSE}
samples.combined <- merge(sample_objs[[1]], y = sample_objs[-1],
                         add.cell.ids =names, project = "CaseStudy2021")
```


```{r}
Idents(object = samples.combined) <- samples.combined@meta.data$state
samples.combined.markers <- Seurat::FindAllMarkers(object = samples.combined)
```

```{r}
# TODO: Use this:

# immune.combined.markers <- markers
# immune.combined.markers %>% filter(avg_log2FC < Inf,
#                                    avg_log2FC > -Inf,
#                                    p_val_adj < 0.05,) %>%
#   group_by(cluster) %>%
#   arrange(desc(avg_log2FC))
```


```{r annotation_biomart}
ensembl <- useEnsembl(biomart = "ensembl", dataset = "mmusculus_gene_ensembl")
genes <- unique(samples.combined.markers$gene)
genedesc <- getBM(attributes=c('external_gene_name',
                               'description',
                               'chromosome_name',
                               'start_position',
                               'end_position'), 
                  filters = 'external_gene_name', 
                  values = genes,
                  mart =ensembl)

samples.combined.markers <- merge(samples.combined.markers, genedesc, all=TRUE,
                                 by.x = c("gene"), by.y = c("external_gene_name"))

# Reorder columns
samples.combined.markers <- samples.combined.markers[c("gene", "cluster", "avg_log2FC", "p_val_adj", "pct.1", "pct.2", "description", "chromosome_name", "start_position", "end_position")]
```


```{r write_excel}
genes_of_note <- c("Sell", "Adam17", "Klf2", "Lef1", "Foxo1", "Sp1", "Ets1",
                   "Mzf1", "Irf1", "Ccr7", "S1pr1", "Cd44")

excel <- createWorkbook("scRNA-seq.xlsx")
addWorksheet(excel, "scRNA-seq")
writeData(excel, sheet = 1, samples.combined.markers)

# Highlight genes of particular interest
samples.combined.markers$rownum <- 1:nrow(samples.combined.markers)
rw <- samples.combined.markers %>% filter(gene %in% genes_of_note )
highlight_style <- openxlsx::createStyle(fgFill = "yellow")
openxlsx::addStyle(wb = excel, 
                   sheet = 1, 
                   style = highlight_style,
                   rows = (rw$rownum+1),
                   cols = 1:ncol(rw),
                   stack = TRUE,
                   gridExpand = TRUE)

# Create second sheet containing only genes of note
rw$rownum <- NULL
rw <- rw %>% arrange(gene)
addWorksheet(excel, "scRNA-seq-specific")
writeData(excel, sheet=2, rw)

# Finally write out to file
saveWorkbook(excel, file = "../results/differential_analysis", overwrite = TRUE)
```


```{r compute_qc_metrics, echo=FALSE}
temp <- c()
for(seurat_obj in sample_objs){
  mito.genes <- rownames(seurat_obj)[grep("^mt-",rownames(seurat_obj))]
  c <- GetAssayData(object = seurat_obj, slot = "counts")
  percent.mito <- colSums(c[mito.genes,])/Matrix::colSums(c)*100
  seurat_obj$percent.mito <- percent.mito
  temp <- append(temp, seurat_obj)
}
sample_objs <- temp
rm(temp)
```


```{r results_setup, echo=FALSE}
if(!dir.exists("../results")){dir.create("../results")}
dir.create("../results/qc_metrics")
```


```{r plot_qc_metrics, echo=FALSE}
library(patchwork)
plots <- list()
for (sample_obj in sample_objs){
  plt <- VlnPlot(object = sample_obj,
                 features = c("nFeature_RNA", "nCount_RNA", "percent.mito"),
                 ncol = 3,
                 combine = FALSE)
  plots <- append(plots, plt)
}
plt <- wrap_plots(plots = plots, nrow = length(sample_objs), ncol = 3)
save_plot("../results/qc_metrics/qc_metrics.png", plt, base_height = 15, base_width = 20)
plt
```


```{r filter_on_qc_metrics, echo=FALSE}
sample_objs <- lapply(X = sample_objs, FUN = function(x) {
  x <- subset(x = x,
              subset = nFeature_RNA > 200 & nFeature_RNA < 2500 & percent.mito < 5 )
})
```


```{r violinplot, echo=FALSE}
plots <- VlnPlot(samples.combined, features = c("Sell"), pt.size = 0, combine = FALSE)
plt <- wrap_plots(plots = plots, nrow = 1)
save_plot("../results/exp_violin.png", plt)
plt
```


```{r featureplot, echo=FALSE}
plt <- FeaturePlot(samples.combined, features = c("Sell"), split.by = "state", 
                   cols = c("grey", "red"))
save_plot("../results/featureplot.png", plt)
```


```{r heatmap, echo=FALSE}
plt <- DoHeatmap(samples.combined, features=c('Sell'))
save_plot("../results/heatmap.png", plt)
a
```


```{r dotplot, echo=FALSE}
plt <- DotPlot(samples.combined, features = rev(c("Sell")), dot.scale = 8) + RotatedAxis()
save_plot("../results/dotplot.png", plt)
plt
```
