---
title: "scRNA-seq Analysis"
author: "Benedict"
date: "17/05/2021"
output:
  pdf_document:
    latex_engine: xelatex
---

# Single Cell RNA-seq Analysis with Seurat
This R Markdown document performs differential gene expression testing for a group of samples and consists of six stages:

1. Data Import
2. Quality Control
3. Dimensionality Reduction
4. Miscellaneous plots
5. Differential Expression (DE) analysis
6. Gene Enrichment Analysis

## Project Structure
```
project
│   README.md
└───data
│   │
│   └───RNA
│   │     N2
│   │     B2
│   │     E2
│   │     M2
│   └───ATAC
│   │     X31-OVA_DIV0
│   │     X31-OVA_DIV3
└───code
│     differential_scRNA-seq.Rmd
│     differential_scATAC-seq.Rmd
└───results
│   └───RNA
│   │   │ scRNA-seq-DE.xlsx
│   │   └───QC
│   │   │ 
│   │   └───plots
│   └───ATAC
│   │   │  scATAC-seq-DE.xlsx
│   │   └───QC
│   │   │ 
│   │   └───plots
```

## Dependencies
* tidyverse
* Seurat
* cowplot
* biomart
* openxlsx
* patchwork
* enrichR
* reshape2

```{r setup, include=FALSE, echo=TRUE, results='hide'}
knitr::opts_chunk$set(echo = TRUE)
setwd(paste0("C:/Users/bened/AppData/Local/Packages/CanonicalGroupLimited.Ubun",
             "tuonWindows_79rhkp1fndgsc/LocalState/rootfs/home/benedict/CaseStu",
             "dy/code"))

# Load Dependencies
packages <- c("tidyverse",  "Seurat", "cowplot", "biomaRt", "openxlsx",
              "patchwork", "enrichR", "reshape2")
#install.packages(packages)
lapply(packages, library, character.only = TRUE)
```

```{r settings}
root_dir <- paste0("C:/Users/bened/AppData/Local/Packages/CanonicalGroupLimited",
                  ".UbuntuonWindows_79rhkp1fndgsc/LocalState/rootfs/home/benedict",
                  "/CaseStudy/data/united/RNA/")
sample_names <- c('X31-OVA-D0', 'X31-OVA-D1', 'X31-OVA-D2', 'X31-OVA-D3', 'X31-OVA-D4')
remove_ribo <-  FALSE
remove_mito <- FALSE
```

## Data Import
```{r data_import,  include=TRUE, echo=TRUE, results='hide'}
samples <- c()

for(sample_name in sample_names){
  sample <- Read10X(data.dir = paste0(root_dir, sample_name, '/filtered_feature_bc_matrix'))
  samples <- append(samples, sample)
}
```

Number of cells per sample pre-QC
```{r pre_qc_cell_counts}
for(i in seq_along(samples)){
  num_cells <- ncol(samples[[i]])
  print(paste0(sample_names[[i]],": ", num_cells))
}
```

## Create Seurat objects from input data
```{r create_seurat_objects, include=TRUE, echo=TRUE}
sample_objs <- c()
for(i in seq_along(samples)){
  sample <- CreateSeuratObject(counts = samples[[i]],
                               project = "scRNA-seq",
                               min.cells = 3,
                               min.features = 200)
  sample@meta.data$state <- sample_names[[i]]
  sample_objs <- append(sample_objs, sample)
  
  num_cells <- ncol(samples[[i]])
  print(paste0(sample_names[[i]],": ", num_cells))
 }
```

\newpage

## Compute QC metrics

Workflow computes and applied the following commonly used metrics in a QC step:

* Number of unique genes detected in each cell
  * Low-quality cells or empty droplets often have very few genes
  * Cell multiplets may have an abnormally high gene count
* Total number of molecules detected within a cell (should correlate with first metric)
*  Percentage of reads mapping to the mitochondrial genome
  * Low-quality and dying cells often contain high mitochondrial contamination
  
```{r compute_qc_metrics, include=TRUE, echo=TRUE, results='hide'}
temp <- c()
for(seurat_obj in sample_objs){
  total_counts_per_cell <- colSums(seurat_obj@assays$RNA@counts)
  
  mito.genes <- rownames(seurat_obj)[grep("^mt-", rownames(seurat_obj))]
  seurat_obj$percent.mito <- (colSums(
    seurat_obj@assays$RNA@counts[mito.genes, ])/total_counts_per_cell)*100
  
  ribo.genes <- rownames(seurat_obj)[grep("^Rp[sl]", rownames(seurat_obj))]
  seurat_obj$percent.ribo <- (colSums(
    seurat_obj@assays$RNA@counts[ribo.genes, ])/total_counts_per_cell)*100
  
  temp <- append(temp, seurat_obj)
}
sample_objs <- temp
rm(temp)
```


```{r plot_qc_metrics, include=TRUE}
# Temp merge for plot purposes only
samples.combined <- merge(sample_objs[[1]], y = sample_objs[-1],
                          add.cell.ids =sample_names, project = "scRNA-seq")
plt_pct_mito <- VlnPlot(object = samples.combined,
                        features = c("percent.mito"),
                        group.by = "state") + labs(title = "% Mitochondrial Reads")
save_plot(paste0("../results/RNA/QC/", "pct_mito", ".png"), plt_pct_mito)


plt_num_features <- VlnPlot(object = samples.combined,
                            features = c("nFeature_RNA"),
                            group.by = "state") + labs(title = "Number of Unique Genes")
save_plot(paste0("../results/RNA/QC/", "num_unique_genes", ".png"), plt_num_features)

# Render
plt_pct_mito
plt_num_features
```

## Filter Genes
Mitochondrial and ribosomal genes are removed
Comment out this code block to keep them
```{r filter_genes, include=FALSE, echo=TRUE}
temp <- c()
for (sample_obj in sample_objs){
# Filter MALAT1
sample_obj <- sample_obj[!grepl("Malat1", rownames(sample_obj)), ]

# Filter Mitocondrial
sample_obj <- sample_obj[!grepl("^mt-", rownames(sample_obj)), ]

# Filter Ribossomal gene (optional if that is a problem on your data) data.filt
sample_obj<- sample_obj[ ! grepl('^Rp[sl]', rownames(sample_obj)), ]
temp <- append(temp, sample_obj)
}

sample_objs <- temp
rm(temp)
```

## Filter cells based on QC metrics
```{r filter_on_qc_metrics, echo=FALSE}
# Set cutoffs based on QC plots
min_features <- 6500
max_mito <- 5

sample_objs <- lapply(X = sample_objs, FUN = function(x) {
  x <- subset(x = x, subset = nFeature_RNA < min_features & percent.mito < max_mito )
})
```

Number of cells per sample post-QC
```{r post_qc_cell_counts}
for(i in seq_along(sample_objs)){
  num_cells <- ncol(sample_objs[[i]])
  print(paste0(sample_names[[i]],": ", num_cells))
}
```

```{r remove_unusable_cells_samples}
samples_to_remove <- c('X31-OVA-D1') # Specify any samples to be removed
idxs_to_remove <- match(samples_to_remove, sample_names)
sample_objs <- sample_objs[-idxs_to_remove]
```

## Normalize Samples independently
The data is normalized using a global-scaling normalization method to normalise gene expression levels for each cell by the total expression.
This is multiplied the a scale factor of 1000 and the result is natural-log-transformed.
```{r normalise, include=FALSE, echo=TRUE}
sample_objs <- lapply(X = sample_objs, FUN = function(x) {
    x <- NormalizeData(x, normalization.method="LogNormalize", scale.factor=1000)
    x <- FindVariableFeatures(x, selection.method = "vst", nfeatures = 2000)
})
```

## Integrate Samples
```{r integrate_objects, echo=TRUE, results='hide'}
anchors <- FindIntegrationAnchors(object.list = sample_objs, dims = 1:20)
samples.combined <- IntegrateData(anchorset = anchors, dims = 1:20)
DefaultAssay(samples.combined) <- "integrated"
```

## Perform Single Integrated Analysis
Pre-processing step before dimensional reduction is performed
* Expression of each gene shifted to give mean expression of 0 across cells
* Expression of each gene scaled to give a variance of 1 across cells
  * Prevents domination from highly-expressed genes
```{r echo=TRUE, results='hide'}
samples.combined <- ScaleData(samples.combined, verbose = FALSE)
samples.combined <- RunPCA(samples.combined, npcs = 30, verbose = FALSE)
```

```{r echo=TRUE, results='hide'}
samples.combined <- RunUMAP(samples.combined, reduction = "pca", dims = 1:30)
samples.combined <- FindNeighbors(samples.combined, reduction = "pca", dims = 1:30)
samples.combined <- FindClusters(samples.combined)
```

```{r visualisation, include=TRUE, echo=TRUE, results='hide'}
plt_umap <- DimPlot(samples.combined, reduction = "umap")
save_plot("umap_clusters.png", plt_umap)
plt_pca <- DimPlot(samples.combined, reduction = "pca")
save_plot("umap_clusters.png", plt_pca)
plt_umap
plt_pca
```

```{r}
genes_to_plot <- c('Sell')

DefaultAssay(samples.combined) <- "RNA"
plt <- FeaturePlot(samples.combined, features = genes_to_plot)
save_plot("genes_umap.png", plt)
plt

```

\newpage

## Perform Differential Expression (DE) Testing
### Compare expression betwen markers
Performs a Wilcoxon Rank Sum test to identify differentially expressed genes between a sample and all other cells, for each sample
```{r}
ensembl <- useEnsembl(biomart = "ensembl", dataset = "mmusculus_gene_ensembl")
```

### Annotate DE genes from ensembl
```{r}
diff_exp <- function(sample_a, sample_b){
  CD8_difference <- Seurat::FindMarkers(object = samples.combined, 
                                        logfc.threshold = 0,
                                        ident.1=sample_a,
                                        ident.2=sample_b,
                                        min.pct=0
                                        )
  genes <- rownames(CD8_difference)
  CD8_difference$gene <- genes
  genedesc <- getBM(attributes=c('external_gene_name',
                                 'description',
                                 'gene_biotype',
                                 'chromosome_name',
                                 'start_position',
                                 'end_position'), 
                    filters = 'external_gene_name', 
                    values = unique(genes),
                    mart = ensembl)

  CD8_difference <- merge(CD8_difference,
                          genedesc, all=TRUE,
                          by.x = c("gene"),
                          by.y = c("external_gene_name"))

  # Reorder & subset columns
  new_order <- c("gene", "avg_log2FC", "p_val_adj", "pct.1", "pct.2",
                 "gene_biotype", "description", "chromosome_name",
                 "start_position", "end_position")
  CD8_difference <- CD8_difference[new_order]
  return(CD8_difference) 
}
```

### Write to excel spreadsheet 'scRNA-seq-DE'
Results of differential gene expression test are written to an excel workbook with two sheets.
Sheet one is all results with genes of particular importance being highlighted in yellow
Sheet two contains only those genes of particular importance

**avg_log2fc: ** log fold-change in average expression between the two groups
**pct.1:** Percentage of cells where gene is detected in first group
**pct.2:** Percentage of cells where gene is detected in second group
**p_val_adj:** Adjusted p-value using bonferroni correction using all genes in dataset

```{r write_excel, include=TRUE, echo=TRUE}
write_to_xlsx <- function(data, filename, genes_of_interest){
  excel <- createWorkbook(filename)
  addWorksheet(excel, "all_genes")
  writeData(excel, sheet = 1, data)
  
  # Highlight genes of particular interest
  data$rownum <- 1:nrow(data)
  rw <- data %>% dplyr::filter(gene %in% genes_of_interest )
  highlight_style <- openxlsx::createStyle(fgFill = "yellow")
  openxlsx::addStyle(wb = excel, 
                     sheet = 1, 
                     style = highlight_style,
                     rows = (rw$rownum+1),
                     cols = 1:ncol(rw),
                     stack = TRUE,
                     gridExpand = TRUE)
  
  # Create second sheet containing only genes of note
  rw$rownum <- NULL
  rw <- rw %>% arrange(gene)
  addWorksheet(excel, "genes_of_interest")
  writeData(excel, sheet=2, rw)
  
  
  # Gene Enrichment Analysis
    # Find and plot top enriched and top depleted GO terms from EnrichR
  go_terms <- enrichr(genes = unique(res$gene),
                      databases = "KEGG_2019_Mouse")[["KEGG_2019_Mouse"]] 
  go_terms <- go_terms[c("Term", "Overlap", "Adjusted.P.value", "Genes")] %>%
    arrange("Overlap")
  addWorksheet(excel, "go_terms")
  writeData(excel, sheet=3, go_terms)
  
  # Finally write out to file
  saveWorkbook(excel, file = paste0("../results/RNA/", filename, ".xlsx"),
               overwrite = TRUE)
  }
```

```{r}
genes_to_show <- c('Sell', 'Klf2', 'Ccr7') # List of genes to plot correlation for

samples.combined <- ScaleData(samples.combined)
matrix <- samples.combined@assays[["RNA"]]@scale.data
matrix_mod <- as.matrix(matrix) 

tmp <- genes_to_show
matrix_mod <- matrix_mod[genes_to_show, ]

for(gene in genes_to_show){
  temp <-as.numeric(matrix_mod[gene,])
  correlations <- apply(matrix_mod,1,function(x){cor(temp,x)})
  tmp <- rbind(tmp, correlations)
}

colnames(tmp) <- genes_to_show
cormat <- tmp[-1,]
rownames(cormat) <- genes_to_show
melted_cormat <- melt(cormat)
melted_cormat$value <- as.numeric(melted_cormat$value)
ggplot(data = melted_cormat, aes(x=Var1, y=Var2, fill=value)) + geom_tile()

# Get lower triangle of the correlation matrix
get_lower_tri <- function(cormat){
  cormat[upper.tri(cormat)] <- NA
  return(cormat)
  }

# Get upper triangle of the correlation matrix
get_upper_tri <- function(cormat){
  cormat[lower.tri(cormat)]<- NA
  return(cormat)
  }

upper_tri <- get_upper_tri(cormat)
  
# Melt the correlation matrix
melted_cormat <- melt(upper_tri, na.rm = TRUE)
melted_cormat$value <- as.numeric(melted_cormat$value)

# Plot Heatmap
plt <- ggplot(data = melted_cormat, aes(Var2, Var1, fill = value)) + 
  geom_tile(color = "white") + 
  scale_fill_gradient2(low = "blue", high = "red", mid = "white", midpoint = 0,
                       limit = c(-1,1), space = "Lab", name="Pearson\nCorrelation") + 
  theme_minimal() + 
  theme(axis.title.x = element_blank(), 
        axis.title.y = element_blank(), 
        panel.grid.major = element_blank(), 
        panel.border = element_blank(), 
        panel.background = element_blank(), 
        axis.ticks = element_blank(), 
        axis.text.x = element_text(angle = 45, vjust = 1, size = 12, hjust = 1)) + 
  coord_fixed()
save_plot("correlation.png", plt)
plt
```